cros_gralloc/cros_gralloc_driver.cc

```diff
@@ -230,151 +204,115 @@ cros_gralloc_driver::cros_gralloc_driver()
+		//
+		// * virtio-GPU backed by inter-VM shared-memory (ivshmem): inter-VM screen cast use
+		//   case. This kind doesn't support importing external buffers neither, and it's
+		//   needed only when the buffers shall be shared for casting.
+		int gpu_grp_type_idx = get_gpu_type(fd);
+
+		if (gpu_grp_type_idx != -1 &&
+		    !(gpu_grp_type_ & (1ull << gpu_grp_type_idx))) {
+			gpu_grp_type_ |= (1ull << gpu_grp_type_idx);
+			driver_fds[gpu_grp_type_idx] = fd;
+		} else if (fd != fallback_fd) {
+			close(fd);
 		}
-
-		node_fd[availabe_node] = fd;
-		int len = snprintf(NULL, 0, "%s", version->name);
-		node_name[availabe_node] = (char *)malloc(len + 1);
-		strncpy(node_name[availabe_node], version->name, len + 1);
-		availabe_node++;
-
 		drmFreeVersion(version);
 	}
 
-	drv_num_ = DRM_RENDER_NODE_START + availabe_node;
-
-	if (availabe_node > 0) {
-		if ((renderer_idx != -1) && (video_idx != -1) && (video_idx != renderer_idx))
-			gpu_grp_type_ |= GPU_TYPE_DUAL_IGPU_DGPU;
-		// if no i915 node found
-		if (renderer_idx == -1) {
-			if (virtio_node_idx != -1) {
-				// Fallback to virtio-GPU
-				video_idx = renderer_idx = virtio_node_idx;
-			} else {
-				drv_loge("Weird scenario! Neither of i915 nor virtio-GPU device is"
-					" found. Use the first deivice.\n");
-				video_idx = renderer_idx = 0;
-			}
-		}
-
-		// Create drv
-		if (!(drv_render_ = drv_create(node_fd[renderer_idx]))) {
-			drv_loge("Failed to create driver for the render device with card id %d\n",
-				renderer_idx);
-			close(node_fd[renderer_idx]);
+restart:
+	for (int i = 0; i < GPU_GRP_TYPE_NR; ++i) {
+		if (!(gpu_grp_type_ & (1ull << i)))
+			continue;
+		if (drivers_[i] != nullptr) {
+			DRV_DESTROY(drivers_[i]);
 		}
-		if (video_idx != renderer_idx) {
-			drv_video_ = drv_create(node_fd[video_idx]);
-			if (!drv_video_) {
-				drv_loge("Failed to create driver for the video device with card id %d\n",
-					video_idx);
-				close(node_fd[video_idx]);
-			}
-		} else {
-			drv_video_ = drv_render_;
-			if (!drv_video_)
-				drv_loge("Failed to create driver for the video device with card id %d\n",
-					video_idx);
+		if (driver_fds[i] == fallback_fd)
+			fallback_fd = -1;
+		struct driver *drv = drv_create(driver_fds[i], gpu_grp_type_);
+		if (!drv) {
+			drv_loge("failed to init minigbm driver on device of type %d\n", i);
+			close(driver_fds[i]);
+			driver_fds[i] = -1;
+			gpu_grp_type_ &= ~(1ull << i);
+			goto restart;
 		}
+		drivers_[i] = drv;
+	}
 
-		if ((virtio_node_idx != -1) && (virtio_node_idx != renderer_idx)) {
-			drv_kms_ = drv_create(node_fd[virtio_node_idx]);
-			if (!drv_kms_) {
-				drv_loge("Failed to create driver for the virtio-gpu device with card id %d\n",
-                                          virtio_node_idx);
-				close(node_fd[virtio_node_idx]);
-				drv_kms_ = drv_render_;
-                        }
-		} else {
-			drv_kms_ = drv_render_;
-		}
+	if (fallback_fd >= 0)
+		close(fallback_fd);
```

Comment:
This 2 lines should put after the next code block "if (gpu_grp_type_ == 0)...", which still use 'fallback_fd'.

Reply:
> This 2 lines should put after the next code block "if (gpu_grp_type_ == 0)...", which still use 'fallback_fd'.

Good catch!